src/System.Text.Json/src/System/Text/Json/Utf8JsonReader.cs

```diff
@@ -0,0 +1,1826 @@
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private void EnsureAndPushStack(JsonTokenType tokenType)
+        {
+            Debug.Assert(tokenType != JsonTokenType.Comment);
+            if (_stack == null)
+            {
+                _stack = new Stack<JsonTokenType>();
+            }
+            _stack.Push(tokenType);
+        }
+
+        private void EndObject()
+        {
+            if (!_inObject || _currentDepth <= 0)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.MismatchedObjectArray, JsonConstants.CloseBrace);
+
+            _consumed++;
+            _bytePositionInLine++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+            {
+                _stackFreeContainer >>= 1;
+                _inObject = (_stackFreeContainer & 1) != 0;
+            }
+            else
+            {
+                Debug.Assert(_stack.Count > 0);
+                _stack.Pop();
+                if (_stack.Count == 0)
+                {
+                    _inObject = (_stackFreeContainer & 1) != 0;
+                }
+                else
+                {
+                    _inObject = _stack.Peek() != JsonTokenType.StartArray;
+                }
+            }
+
+            _currentDepth--;
+            _tokenType = JsonTokenType.EndObject;
+        }
+
+        private void StartArray()
+        {
+            if (_currentDepth >= _maxDepth)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ArrayDepthTooLarge);
+
+            _currentDepth++;
+            _consumed++;
+            _bytePositionInLine++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+            {
+                _stackFreeContainer = _stackFreeContainer << 1;
+            }
+            else
+            {
+                EnsureAndPushStack(JsonTokenType.StartArray);
+            }
+
+            _tokenType = JsonTokenType.StartArray;
+            _inObject = false;
+        }
+
+        private void EndArray()
+        {
+            if (_inObject || _currentDepth <= 0)
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.MismatchedObjectArray, JsonConstants.CloseBracket);
+
+            _consumed++;
+            _bytePositionInLine++;
+
+            if (_currentDepth <= JsonReaderState.StackFreeMaxDepth)
+            {
+                _stackFreeContainer >>= 1;
+                _inObject = (_stackFreeContainer & 1) != 0;
+            }
+            else
+            {
+                Debug.Assert(_stack.Count > 0);
+                _stack.Pop();
+                if (_stack.Count == 0)
+                {
+                    _inObject = (_stackFreeContainer & 1) != 0;
+                }
+                else
+                {
+                    _inObject = _stack.Peek() != JsonTokenType.StartArray;
+                }
+            }
+
+            _currentDepth--;
```

Comment:
All of the code from just after the opening ThrowHelper call through this line appear to be exact duplicates between EndObject and EndArray.  Can that all be separated out into a helper to avoid the code duplication?